debezium-server-redis/src/main/java/io/debezium/server/redis/RedisMemoryThreshold.java

```diff
@@ -26,55 +25,79 @@ public class RedisMemoryThreshold {
+     * it is impossible to rely on delayed memory usage reading to prevent OOM.
+     * In order to protect the Redis database and throttle down the sink we have created this mechanism:...
+     * @param extraMemory   - Estimated size of a single record.
+     * @param bufferSize    - Number of records in a batch.
+     * @param bufferFillRate - Rate in which memory can be filled.
+     * @return
+     */
+    public boolean checkMemory(long extraMemory, int bufferSize, int bufferFillRate) {
+        Tuple2<Long, Long> memoryTuple = memoryTuple();
+
+        maximumMemory = memoryTuple.getItem2();
+
+        if (maximumMemory == 0) {
+            totalProcessed += bufferSize;
+            LOGGER.debug("Total Processed Records: {}", totalProcessed);
             return true;
         }
-        long maxMemory = memoryTuple.getItem2();
-        if (maxMemory > 0) {
-            long usedMemory = memoryTuple.getItem1();
-            long percentage = usedMemory * 100 / maxMemory;
-            if (percentage >= memoryThreshold) {
-                LOGGER.info("Sink memory threshold percentage was reached. Will retry; " +
-                        "(current: {}%, configured: {}%, used_memory: {}, maxmemory: {}).",
-                        percentage, memoryThreshold,
-                        usedMemory, maxMemory);
-                return false;
-            }
+
+        long extimatedBatchSize = extraMemory * bufferFillRate;
+        long usedMemory = memoryTuple.getItem1();
+        long prevAccumulatedMemory = accumulatedMemory;
+        long diff = usedMemory - previouslyUsedMemory;
+        if (diff == 0L) {
+            accumulatedMemory += extraMemory * bufferSize;
+        }
+        else {
+            previouslyUsedMemory = usedMemory;
+            accumulatedMemory = extraMemory * bufferSize;
+        }
+        long estimatedUsedMemory = usedMemory + accumulatedMemory + extimatedBatchSize;
+
+        if (estimatedUsedMemory >= maximumMemory) {
+            LOGGER.info(
+                    "Sink memory threshold percentage was reached. Will retry; "
+                            + "(estimated used memory size: {}, maxmemory: {}). Total Processed Records: {}",
+                    humanReadableSize(estimatedUsedMemory), humanReadableSize(maximumMemory), totalProcessed);
+            accumulatedMemory = prevAccumulatedMemory;
+            return false;
+        }
+        else {
+            LOGGER.debug(
+                    "Maximum reached: {}; Used Mem {}; Max Mem: {}; Accumulate Mem: {}; Estimated Used Mem: {}, Record Size: {}, NumRecInBuff: {}; Total Processed Records: {}",
+                    (estimatedUsedMemory >= maximumMemory), humanReadableSize(usedMemory), humanReadableSize(maximumMemory),
+                    humanReadableSize(accumulatedMemory), humanReadableSize(estimatedUsedMemory),
+                    humanReadableSize(extraMemory), bufferSize, totalProcessed + bufferSize);
+            totalProcessed += bufferSize;
```

Comment:
Same as above, could this eventually overflow?

Reply:
Same answer as above